# Data Visualization & AI System


## Dataset: YouTube Video Analytics


## Data Preprocessing


In [49]:
import pandas as pd
import numpy as np

In [50]:
df = pd.read_csv("youtube_data.csv")

In [51]:
df.shape

(17589, 17)

In [52]:
df.head()

video_id  duration  bitrate  bitrate(video)  height  width  frame rate  \
0  --F7dc-_FSI       180     5777            5640    1920   1080       25.00   
1  --cCAD-8Y_U       930     1195            1001    1280    720       30.00   
2  --g2gG8pQ0w       233     3028            2833    1280    720       23.98   
3  -0DR7-voRCU       562      431             300     320    240       29.97   
4  -0Fkp-2EzX0       300     3087            2929    1280    720       23.98   

   frame rate(est.) codec         category  \
0             25.00  h264  News & Politics   
1             30.00  h264           Gaming   
2             23.98  h264           Comedy   
3              0.00  h264   People & Blogs   
4             23.98  h264           Gaming   

                                           url  \
0  https://www.youtube.com/watch?v=--F7dc-_FSI   
1  https://www.youtube.com/watch?v=--cCAD-8Y_U   
2  https://www.youtube.com/watch?v=--g2gG8pQ0w   
3  https://www.youtube.com/watch?v=-0DR7-voRCU   
4  https://www.youtube.com/watch?v=-0Fkp-2EzX0   

                                               title  \
0              «السودان ينتفض» أمام السفارة بالقاهرة   
1                 Pokemon Tower Defense   Episodio 2   
2               New Hip Hop - Kemo Treats - Pancakes   
3                         Sister's Keeper Impression   
4  JBHAMMER777 (AKUMA) VS KtossPol  Rank 1 (Vega)...   

                                         description  \
0  حرية سلام وعدالة، الثورة خيار الشعب.. بهذا اله...   
1                Espero que tenham gostado do video!   
2  Download all our music for FREE just enter $0....   
3  A Review of the Shampoo Bar(Sunburst) and Dip ...   
4  JBHAMMER777 is he strong enough to fight again...   

                                            hashtags  views  likes  comments  
0  في, عسكري, حكم, حرية, بخصوص, المصري اليوم, الس...   4078     41         3  
1  Pokémon (Anime/Manga Franchise), Pokémon (Vide...     33      2         1  
2  breaking bad spoof, Recipes, Pancakes, canada,...  13227    135        17  
3  natural hair, first impression, sister's keepe...     43      1         0  
4  Tokido, momochi, (Ryu), AE, 11, at, Piyoppia, ...    222      2         0

In [53]:
df.tail()

video_id  duration  bitrate  bitrate(video)  height  width  \
17584  zyK7-WQTDaY       571     1285            1175     640    480   
17585  zyPO-hxvWDg       105      840             719     640    480   
17586  zyYk-W_JqNY       161      920             820     640    480   
17587  zz3a-wnxBrs       291     4892            4738    1920   1080   
17588  zzF2-Z2B5Wc       244      392             302     320    240   

       frame rate  frame rate(est.) codec        category  \
17584       29.92              0.00  h264  People & Blogs   
17585       29.97              0.00  h264  People & Blogs   
17586       29.97              0.00  h264  People & Blogs   
17587       29.97             29.97  h264          Gaming   
17588       25.00              0.00  h264  People & Blogs   

                                               url  \
17584  https://www.youtube.com/watch?v=zyK7-WQTDaY   
17585  https://www.youtube.com/watch?v=zyPO-hxvWDg   
17586  https://www.youtube.com/watch?v=zyYk-W_JqNY   
17587  https://www.youtube.com/watch?v=zz3a-wnxBrs   
17588  https://www.youtube.com/watch?v=zzF2-Z2B5Wc   

                                                   title  \
17584                             rentero 1ª parte 2.MOD   
17585  Cascavel faz campanha impecável em campeonato ...   
17586  Marcha para Cristo.San Francisco Menéndez.  5 ...   
17587                             DK Solo Steinerne Kern   
17588  &#x202b;يارب نور دربي لعبد المجيد الفوزان بجود...   

                                             description  \
17584                                                      
17585                                                      
17586                                                NaN   
17587  Das ist mein erstes Video, also ist noch alles...   
17588                                                      

                                                hashtags  views  likes  \
17584                   rentero, latapialinares, iu, pca     27      0   
17585                                                NaN    295      1   
17586                               Cristo, Marcha, para    770      3   
17587  Steinerne Kern, Vitriolsteindrachen, DK Solo, ...     42      1   
17588  نور, بجوده, دربي, &#x202b;يارب, المجيد, عاليه&...     20      0   

       comments  
17584         0  
17585         0  
17586         0  
17587         0  
17588         0

In [54]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17589 entries, 0 to 17588
Data columns (total 17 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   video_id          17589 non-null  object 
 1   duration          17589 non-null  int64  
 2   bitrate           17589 non-null  int64  
 3   bitrate(video)    17589 non-null  int64  
 4   height            17589 non-null  int64  
 5   width             17589 non-null  int64  
 6   frame rate        17589 non-null  float64
 7   frame rate(est.)  17589 non-null  float64
 8   codec             17589 non-null  object 
 9   category          17589 non-null  object 
 10  url               17589 non-null  object 
 11  title             17589 non-null  object 
 12  description       16477 non-null  object 
 13  hashtags          16013 non-null  object 
 14  views             17589 non-null  int64  
 15  likes             17589 non-null  int64  
 16  comments          17589 non-null  int64 

In [55]:
df.isnull().sum()

video_id               0
duration               0
bitrate                0
bitrate(video)         0
height                 0
width                  0
frame rate             0
frame rate(est.)       0
codec                  0
category               0
url                    0
title                  0
description         1112
hashtags            1576
views                  0
likes                  0
comments               0
dtype: int64

In [56]:
df.duplicated().sum()

np.int64(0)

In [57]:
df.describe()

duration       bitrate  bitrate(video)        height         width  \
count  17589.000000  17589.000000    17589.000000  17589.000000  17589.000000   
mean     241.551936   1271.354369     1150.418443    766.781170    504.591961   
std      493.026994   1375.359875     1351.800202    467.289304    262.727746   
min        1.000000      0.000000        0.000000    108.000000     88.000000   
25%       51.000000    437.000000      326.000000    426.000000    320.000000   
50%      135.000000    743.000000      632.000000    640.000000    480.000000   
75%      268.000000   1293.000000     1184.000000    960.000000    720.000000   
max    25845.000000  22421.000000    22229.000000   2592.000000   1944.000000   

         frame rate  frame rate(est.)         views          likes  \
count  17589.000000      17589.000000  1.758900e+04   17589.000000   
mean      26.467639          9.471172  6.894158e+04     208.862641   
std        6.039748         13.253197  3.537491e+06    8477.504735   
min        3.750000          0.000000  0.000000e+00       0.000000   
25%       25.000000          0.000000  3.800000e+01       0.000000   
50%       29.970000          0.000000  1.520000e+02       1.000000   
75%       29.970000         25.000000  8.000000e+02       4.000000   
max       59.080000         30.000000  4.103849e+08  836981.000000   

           comments  
count  17589.000000  
mean      12.899312  
std      225.839378  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        1.000000  
max    16634.000000

In [58]:
(df["views"] == 0).sum(), (df["likes"] == 0).sum(), (df["comments"] == 0).sum()

(np.int64(171), np.int64(7918), np.int64(12899))

In [59]:
numeric_features = ["duration", "bitrate", "height", "width", "frame rate", "views", "likes", "comments"]
(df[numeric_features] < 0).sum()

duration      0
bitrate       0
height        0
width         0
frame rate    0
views         0
likes         0
comments      0
dtype: int64

### Initial Observations

- The dataset contains **17,589 rows and 17 columns**, which is below the required minimum of 20,000 rows for this project. Therefore data augmentation will be applied to meet the requirement.
- The dataset includes a mix of **numerical features** and **categorical/text features**.
- Missing values are present only in textual columns:
  - `description` contains 1,112 missing values
  - `hashtags` contains 1,576 missing values
- No duplicate rows were detected in the dataset.
- All numerical features are complete and stored in appropriate numeric data types.


## Feature Engineering

In [60]:
df["has_description"] = df["description"].notna().astype(int)
df["has_hashtags"] = df["hashtags"].notna().astype(int)

df = df.drop(columns=["description", "hashtags"])

In [61]:
df.isnull().sum()

video_id            0
duration            0
bitrate             0
bitrate(video)      0
height              0
width               0
frame rate          0
frame rate(est.)    0
codec               0
category            0
url                 0
title               0
views               0
likes               0
comments            0
has_description     0
has_hashtags        0
dtype: int64

In [62]:
df["like_rate"] = df["likes"] / (df["views"] + 1)
df["comment_rate"] = df["comments"] / (df["views"] + 1)

In [63]:
df["is_hd"] = (df["height"] >= 720).astype(int)

In [64]:
df["short_vid"] = (df["duration"] < 1200).astype(int)        # < 20 mim
df["long_vid"] = (df["duration"] >= 1200).astype(int)        # > 20 min

In [65]:
df = df.drop(columns=["video_id", "url", "title"])

In [66]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17589 entries, 0 to 17588
Data columns (total 19 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   duration          17589 non-null  int64  
 1   bitrate           17589 non-null  int64  
 2   bitrate(video)    17589 non-null  int64  
 3   height            17589 non-null  int64  
 4   width             17589 non-null  int64  
 5   frame rate        17589 non-null  float64
 6   frame rate(est.)  17589 non-null  float64
 7   codec             17589 non-null  object 
 8   category          17589 non-null  object 
 9   views             17589 non-null  int64  
 10  likes             17589 non-null  int64  
 11  comments          17589 non-null  int64  
 12  has_description   17589 non-null  int64  
 13  has_hashtags      17589 non-null  int64  
 14  like_rate         17589 non-null  float64
 15  comment_rate      17589 non-null  float64
 16  is_hd             17589 non-null  int64 

In [67]:
df = pd.get_dummies(df, columns=["codec"], drop_first=True)

In [68]:
df = pd.get_dummies(df, columns=["category"], drop_first=True)

In [69]:
df["popularity_score"] = (0.5*np.log1p(df["views"]) + 0.3*np.log1p(df["likes"]) + 0.2*np.log1p(df["comments"]))

In [70]:
df.columns

Index(['duration', 'bitrate', 'bitrate(video)', 'height', 'width',
       'frame rate', 'frame rate(est.)', 'views', 'likes', 'comments',
       'has_description', 'has_hashtags', 'like_rate', 'comment_rate', 'is_hd',
       'short_vid', 'long_vid', 'codec_h264', 'codec_mpeg4', 'codec_vp8',
       'category_Comedy', 'category_Education', 'category_Entertainment',
       'category_Film & Animation', 'category_Gaming',
       'category_Howto & Style', 'category_Music', 'category_News & Politics',
       'category_Nonprofits & Activis', 'category_People & Blogs',
       'category_Pets & Animals', 'category_Science & Technology',
       'category_Shows', 'category_Sports', 'category_Travel & Events',
       'popularity_score'],
      dtype='object')

## Data Augmentation

In [74]:
continuous_features = ["duration", "bitrate", "bitrate(video)", "height", "width", "frame rate",
                       "frame rate(est.)", "like_rate", "comment_rate", "popularity_score" ]


In [75]:
augmented_df = df.sample(n=5000, replace=True, random_state=42).copy()


In [76]:
for col in continuous_features:
    noise = np.random.normal( loc=0, scale=0.01 * augmented_df[col].std(), size=augmented_df.shape[0] )
    augmented_df[col] = augmented_df[col] + noise

In [77]:
augmented_df[continuous_features] = augmented_df[continuous_features].clip(lower=0)

In [83]:
augmented_df.shape

(5000, 36)

In [84]:
augmented_df.head()

duration      bitrate  bitrate(video)       height        width  \
15795  142.108317   460.729118      364.833213   351.823969   290.065394   
860     45.287670   477.828889      324.838681  1287.378140   477.542157   
5390     2.308386   747.154010      631.298653   481.544782   357.377788   
11964  352.488759  1929.131792     1749.453876  1279.107038   719.439973   
11284  304.651296  6192.282736     6011.888532  1929.111860  1079.607387   

       frame rate  frame rate(est.)  views  likes  comments  ...  \
15795   24.959362          0.074351   1241     36         0  ...   
860     15.023057         14.827653  11052      2         4  ...   
5390    30.025734          0.000000     25      0         0  ...   
11964   29.929503         29.933558   2076      4         0  ...   
11284   29.935420         29.827937    214      0         0  ...   

       category_Music  category_News & Politics  \
15795           False                      True   
860             False                     False   
5390            False                     False   
11964           False                     False   
11284           False                     False   

       category_Nonprofits & Activis  category_People & Blogs  \
15795                          False                    False   
860                            False                    False   
5390                           False                    False   
11964                          False                    False   
11284                          False                    False   

       category_Pets & Animals  category_Science & Technology  category_Shows  \
15795                    False                          False           False   
860                      False                           True           False   
5390                     False                          False           False   
11964                    False                          False           False   
11284                    False                          False           False   

       category_Sports  category_Travel & Events  popularity_score  
15795            False                     False          4.632145  
860              False                     False          5.321925  
5390             False                     False          1.607371  
11964            False                     False          4.325648  
11284            False                     False          2.698603  

[5 rows x 36 columns]

In [85]:
augmented_df.tail()

duration      bitrate  bitrate(video)       height       width  \
5480  241.576575  2119.199631     2010.506604  1283.959663  715.934258   
8033  209.350523   700.597618      586.271138   638.055140  475.812273   
983   188.389657   367.122858      284.084660   324.004586  233.671963   
8543   42.812807  1290.562195     1205.159675   643.047872  478.150370   
9916  281.059007   893.799747      756.774346   642.977342  355.822011   

      frame rate  frame rate(est.)  views  likes  comments  ...  \
5480   30.017258         30.072356    228      4         0  ...   
8033   30.082408          0.000000   6251     31         8  ...   
983    30.036915          0.004005     44      0         0  ...   
8543   29.994322          0.032675    183      1         0  ...   
9916   29.889673          0.000000   1571     10         1  ...   

      category_Music  category_News & Politics  category_Nonprofits & Activis  \
5480            True                     False                          False   
8033            True                     False                          False   
983             True                     False                          False   
8543            True                     False                          False   
9916            True                     False                          False   

      category_People & Blogs  category_Pets & Animals  \
5480                    False                    False   
8033                    False                    False   
983                     False                    False   
8543                    False                    False   
9916                    False                    False   

      category_Science & Technology  category_Shows  category_Sports  \
5480                          False           False            False   
8033                          False           False            False   
983                           False           False            False   
8543                          False           False            False   
9916                          False           False            False   

      category_Travel & Events  popularity_score  
5480                     False          3.178765  
8033                     False          5.855656  
983                      False          1.925243  
8543                     False          2.817965  
9916                     False          4.537412  

[5 rows x 36 columns]

In [78]:
df_augmented = pd.concat([df, augmented_df], ignore_index=True)

In [79]:
df.shape, df_augmented.shape

((17589, 36), (22589, 36))

In [80]:
df["popularity_score"].describe(), df_augmented["popularity_score"].describe()

(count    17589.000000
 mean         3.099106
 std          1.806232
 min          0.000000
 25%          1.903331
 50%          2.716861
 75%          3.841463
 max         15.183520
 Name: popularity_score, dtype: float64,
 count    22589.000000
 mean         3.099720
 std          1.801216
 min          0.000000
 25%          1.903331
 50%          2.721209
 75%          3.834938
 max         15.183520
 Name: popularity_score, dtype: float64)

In [82]:
df_augmented.to_csv("youtube_augmented.csv", index=False)

## Exploratory Data Analysis (EDA)